lib/test/lis/types.cpp

```diff
@@ -130,49 +131,312 @@ TEST_CASE("16-bit floating point", "[type]") {
+
+        float v;
+        lis_f32low( input, &v );
+        CHECK( v == expected );
+        CHECK (v > 0);
+    }
+
+    SECTION(" smallest LIS-stored positive number (data loss) ") {
+        /* E = -2^15, M = 1 */
+        const float expected =
+            std::pow(2.0f, -15) * std::pow(2.0f, -std::pow(2.0f, 15) );
+        const bytes<4> input = { 0x80, 0x00, 0x00, 0x01 };
+
+        float v;
+        lis_f32low( input, &v );
+        CHECK( v == expected );
+        // Failing: data is not 0, but due to precision loss is reported as 0
+        // CHECK (v > 0);
     }
 }
 
 TEST_CASE("32-bit fixed point", "[type]") {
+
+    SECTION("standard floats") {
+        const std::array< bytes<4>, 4 > inputs = {{
+            { 0x00, 0x00, 0x00, 0x00 }, // 0
+            { 0x00, 0x00, 0x80, 0x00 }, // 0.5
+            { 0x00, 0x99, 0x40, 0x00 }, // 153.25
+            { 0xFF, 0x66, 0xC0, 0x00 }, // -153.25
+        }};
+
+        const std::array< float, inputs.size() > expected = {
+            0,
+            0.5,
+            153.25,
+            -153.25,
+        };
+
+        for( std::size_t i = 0; i < expected.size(); ++i ) {
+            float v;
+            lis_f32fix( inputs[ i ], &v );
+            CHECK( v == expected[ i ] );
+        }
+    }
+
+    SECTION(" Additional precision loss due to available bits number ") {
+        /* Looks like we lose some precision due to type having 31 significant
+         * digits and float having just 23 places in fraction. So, for example,
+         * 2^9 - 2^(-16) already loses precision.
+         *
+         * This, however, shouldn't be a problem if doubles were used as their
+         * fraction is 52 bits
+         */
+        SECTION(" No precision loss ") {
+            /* 2^8 - 2^(-16) */
+            const float expected1 = std::pow(2.0f, 8) - std::pow(2.0f, -16);
+            const bytes<4> input1 = { 0x00, 0xFF, 0xFF, 0xFF };
+
+            /* 2^8 */
+            const float expected2 = std::pow(2.0f, 8);
+            const bytes<4> input2 = { 0x01, 0x00, 0x00, 0x00 };
+
+            float v1;
+            lis_f32fix( input1, &v1 );
+            CHECK( v1 == expected1 );
+
+            float v2;
+            lis_f32fix( input2, &v2 );
+            CHECK( v2 == expected2 );
+
+            CHECK( v1 != v2 );
+        }
+
+        SECTION(" Precision loss ") {
+            /* 2^9 - 2^(-16) */
+            const float expected1 = std::pow(2.0f, 9) - std::pow(2.0f, -16);
+            const bytes<4> input1 = { 0x01, 0xFF, 0xFF, 0xFF };
+
+            /* 2^9 */
+            const float expected2 = std::pow(2.0f, 9);
+            const bytes<4> input2 = { 0x02, 0x00, 0x00, 0x00 };
+
+            float v1;
+            lis_f32fix( input1, &v1 );
+            CHECK( v1 == expected1 );
+
+            float v2;
+            lis_f32fix( input2, &v2 );
+            CHECK( v2 == expected2 );
+
+            // value can't be represented by float as a different number
```

Comment:
What do you mean? should this stay?

Reply:
That's the other wording for precision loss.
input1 and input2 are different numbers in the original type.
but after conversion they are different due to precision loss due to float limits.